Implement subgraph publishing

AustinMroz opened the pull request:
Adds a system for saving blueprints of subgraphs for reuse later.  When a subgraph is selected, a new publish icon is added to the selection toolbar. Subgraph blueprints can by added by the node library or the node searchbox. When accessed from the library, options are provided for editing or deleting a subgraph.

These blueprints are saved to the user data folder and will persist across different browsers.

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5139-Implement-subgraph-publishing-2566d73d365081b7a51ac6ed2345f215) by [Unito](https://www.unito.io)

AustinMroz committed:
Implement subgraph publishing

AustinMroz committed:
Add missing null check

AustinMroz committed:
Fix subgraph blueprint display in workflows tab

AustinMroz committed:
Fix demotion of subgraph blueprints on reload

Commit:
Update locales [skip ci]

AustinMroz committed:
Update blueprint def on save, cleanup

AustinMroz committed:
Fix skipped tracking on subgraph publish

When a subgraph is first published, it previously was not added to the
subgraphCache. This would cause deletion to fail until a reload occurred.

AustinMroz committed:
Fix failing vite tests

A couple of tests that were mocking classes broke SubgraphBlueprint
inheritance. Since they aren't testing anythign related to subgraph
blueprints, the subgraph store is mocked as well.

AustinMroz committed:
Make blueprint breadcrumb badge clickable

AustinMroz committed:
Add confirmation for overwrite on publish

DrJKL commented on src/lib/litegraph/src/LGraphCanvas.ts:
```diff
@@ -3728,6 +3730,14 @@ export class LGraphCanvas
   _pasteFromClipboard(
     options: IPasteFromClipboardOptions = {}
+  ): ClipboardPasteResult | undefined {
+    const data = localStorage.getItem('litegrapheditor_clipboard')
```
https://vueuse.org/core/useLocalStorage/
You might be able to use the composable instead. Should help prevent typo based key misalignment.

DrJKL replied:
I think it'd also help with the [serialize/deserialize](https://vueuse.org/core/useStorage/#custom-serialization)

DrJKL commented on src/scripts/app.ts:
```diff
@@ -796,6 +797,7 @@ export class ComfyApp {
     this.resizeCanvas()
 
     await useWorkspaceStore().workflow.syncWorkflows()
+    await useSubgraphStore().fetchSubgraphs()
     await useExtensionService().loadExtensions()
```
Do these conflict with one another? If not, could we batch them with [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled)?

AustinMroz replied:
I don't think extensions can access the workflowStore or the subgraphStore, but it's not something I'd want to risk.
On the other hand, I hadn't realized Promise.All was sequential. Will have to fix the incorrect usage of it inside fetchSubgraphs()

DrJKL commented on src/services/litegraphService.ts:
```diff
@@ -948,6 +949,20 @@ export const useLitegraphService = () => {
+      const results = canvas._deserializeItems(strippedItems, {
+        position: options.pos
+      })
+      if (!results) throw new Error('Failed to add subgraph blueprint')
+      const node = results.nodes.values().next().value
+      if (!node) throw new Error('Failed to add subgraph blueprint')
```
It might be helpful to differentiate between the ways this is failing, if a consumer only looks at the message and not the stacktrace.

DrJKL commented on src/services/litegraphService.ts:
```diff
@@ -948,6 +949,20 @@ export const useLitegraphService = () => {
+      const canvas = canvasStore.getCanvas()
+      const bp = useSubgraphStore().getBlueprint(nodeDef.name)
+      const items = { nodes: bp.nodes, subgraphs: bp.definitions?.subgraphs }
+      //FIXME
```
//FIXME

DrJKL commented on src/stores/nodeDefStore.ts:
```diff
@@ -291,8 +292,14 @@ export const useNodeDefStore = defineStore('nodeDef', () => {
-  const nodeDefs = computed(() => Object.values(nodeDefsByName.value))
+  const nodeDefs = computed(() => {
+    return [
+      ...Object.values(nodeDefsByName.value),
```
```suggestion
      ...Object.values(nodeDefsByName),
```

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+      )
+    const subgraphs = this.activeState.definitions.subgraphs
+    const nodes = this.activeState.nodes
+    //Instanceof doesn't funciton as nodes are serialized
```
Could we use Zod to validate?

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+      throw new Error(
+        'The root graph of a subgraph blueprint must consist of only a single subgraph node'
+      )
+    const subgraphs = this.activeState.definitions.subgraphs
+    const nodes = this.activeState.nodes
```
If you're feeling fancy...
```suggestion
    const { subgraphs } = this.activeState.definitions
    const { nodes } = this.activeState
```

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+    this.validateSubgraph()
+    if (!this.hasPromptedSave) {
+      //Swap to saveAs?
+      const newname = await useDialogService().prompt({
```
`newName`?

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+  override async saveAs(path: string) {
+    this.validateSubgraph()
+    this.hasPromptedSave = true
+    useSubgraphStore().updateDef(this as LoadedComfyWorkflow)
```
Why do you need to cast?

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+      await api.listUserDataFullInfo(SubgraphBlueprint.basePath)
+    ).filter((f) => f.path.endsWith('.json'))
+    await Promise.all(
+      res.map(async (f) => {
+        const name = f.path.slice(0, -5)
+        f.path = SubgraphBlueprint.basePath + f.path
+        const bp = await new SubgraphBlueprint(f).load()
+        useWorkflowStore().attachWorkflow(bp)
+        const nodeDef = convertToNodeDef(bp)
+
+        subgraphDefCache.value.set(name, nodeDef)
+        subgraphCache[name] = bp
+      })
```
It might be cleaner to extract a function here.

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+    const subgraphNode = workflow.changeTracker.initialState.nodes[0]
+    if (!subgraphNode) throw new Error('Invalid Subgraph Blueprint')
+    const uuid = subgraphNode.type
+    if (!uuid) throw new Error('')
```
Empty Error message

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+    const inputs = Object.fromEntries(
+      subgraphNode.inputs.map((i) => [
+        i.name,
+        ['' + i.type, undefined] satisfies InputSpec
```
Very nit
```suggestion
        [`${i.type}`, undefined] satisfies InputSpec
```

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+    )
+    const nodedefv1: ComfyNodeDefV1 = {
+      input: { required: inputs },
+      output: subgraphNode.outputs.map((o) => '' + o.type),
```
Same here, unless there's a style guide rule that string casts are done by concatenating `''`

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+      defaultValue: subgraphNode.title
+    })
+    if (!name) return
+    if (subgraphDefCache.value.has(name))
+      if (!(await confirmOverwrite(name)))
```
```suggestion
    if (subgraphDefCache.value.has(name) && !(await confirmOverwrite(name)))
```

DrJKL commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,254 @@
+      throw new Error('not yet loaded')
+    return subgraphCache[name].changeTracker.initialState
+  }
+  function deleteBlueprint(nodeType: string) {
```
Would you mind spacing out the functions in here?

DrJKL commented on src/stores/workflowStore.ts:
```diff
@@ -309,6 +310,20 @@ export const useWorkflowStore = defineStore('workflow', () => {
     const fullPath = getUnconflictedPath(
       ComfyWorkflow.basePath + (path ?? 'Unsaved Workflow.json')
     )
+    //TODO: Reconsider. Referencing subgraph constants outside
```
Agreed.

DrJKL commented on src/stores/workflowStore.ts:
```diff
@@ -357,7 +372,10 @@ export const useWorkflowStore = defineStore('workflow', () => {
 
   const persistedWorkflows = computed(() =>
-    Array.from(workflows.value).filter((workflow) => workflow.isPersisted)
+    Array.from(workflows.value).filter(
```
```suggestion
    Array.from(workflows).filter(
```

DrJKL reviewed (commented):
Looks good, mostly just code style comments 👍🏻

AustinMroz committed:
Simplify blueprint badge naming

AustinMroz committed:
Swap to promise.allSettled when fetching subgraphs

AustinMroz committed:
Navigate into subgraph on blueprint edit

AustinMroz committed:
Revert mission of value in blueprint breadcrumb

This was causing the blueprint badge to always display

AustinMroz committed:
Misc code quality fixes

AustinMroz committed:
Set subgraphNode title on blueprint add.

When a subgraph blueprint is added to the graph, the title of the
subgraphNode is now set to be the title of the blueprint.

NOTE: The name of the subgraph node when a blueprint is edited is left
unchanged. This may cause minor user confusion.

AustinMroz committed:
Add "Delete Blueprint" option to breadcrumb

When editing a blueprint, the options provided for the root graph of the
breadcrumb included a Delete Workflow option. This still functioned for
deleting the current blueprint when selected, but didn't make sense. It
has been updated to instead describe that it deletes the current
blueprint

AustinMroz committed:
Extract subgraph load code as function

AustinMroz committed:
Fix subgraphs appearing in library after refresh

Subgraph nodes were hidden from the node library and context menu by
setting skip_list to true. Unfortunately, this causes them to be
mistakenly be caught and registered as vue nodes when a refresh is
performed. This is fixed by adding a check for skip_list.

AustinMroz committed:
Add delete button and confirmation for deletion

AustinMroz committed:
Use more specific warning for blueprint deletion

AustinMroz committed:
At success toast on subgraph publish

Will return later to potentially add a node library link to the toast

AustinMroz committed:
Don't apply subgraph context menu to normal nodes

Subgraph blueprints have a right click -> delete option in the node
library.  This was incorrectly being dislplayed on non blueprint nodes.

AustinMroz committed:
Remove hardcoded subgraphs path

Rather happy with this change. Rather than trying to introduce a
recursive import to pass a magic string, this solution is both
sufficient AND allows potential future extensions with less breakage.

AustinMroz committed:
Fix nodeDef update on save

Wait to update the node def cache until after a blueprint has been
saved. Before, changes to links weren't actually being made visisble.

AustinMroz committed:
Fix SaveAs with subgraph blueprints

AustinMroz committed:
Remove ugly serialize/deserialize

Thought I had already tested this, and found that the mere existence of
proxies was causing issues, but simply adding a correct annotation is
sufficient now.

AustinMroz committed:
Improve error specificity

AustinMroz committed:
Framework for user defined blueprint descriptions

BlueprintDescription can be added to a workflows extra field to provide
more useful information about a blueprint's purpose

Actually hooking this up in a way that is user accessible is out of
scope for right now, but this will simplify future implementation.

AustinMroz committed:
Cleanup breadcrumb dropdown options

Removes Dupliate for blueprints, adds a publish subgraph option.

The publish subgraph button currently routes through the save as logic.
Unforunately, this results in the prompt for name referencing workflows.
The cleanest way to resolve this is still being considered

AustinMroz committed:
Move blueprint renaming into blueprint load

Blueprints should automatically set the name of the added node to the
filename when added. This mostly worked, but created uglier edgecases:
The subgraph itself wasn't renamed, and it would need to be
reimplemented to apply when editing a blueprint.

Instead, this is now applied when a subgraphBlueprint is first loaded.
This keeps all the logic routed through a single point

AustinMroz committed:
Move saveAs prompt into workflow class

Ensures that the correct publish text is displayed when editing
blueprints without making an awful mess of imports

AustinMroz committed:
Fix tests by making subgraphBlueprint internal

This has the added benefit of forcing better organization.

Reverts the useWorkflowThumbnail patch as it is no longer required.

AustinMroz committed:
Add tests for subgraph blueprints

AustinMroz committed:
Rewrite confirmation dialog

AustinMroz committed:
Fix overwrite on publish new subgraph

1 is used as a placeholder size as -1 indicates the baking userFile is
temporary, not persisted, and therefore, not able to overwrite when
saved.

AustinMroz committed:
When editing blueprint, tint background blue

AustinMroz committed:
Fix blueprint tint at low LOD

AustinMroz committed:
Set node source for blueprints to Blueprint

AustinMroz committed:
Fix publish test

Making subgraph blueprints non temporary on publish made it so the
following load actually occurs. A mock has been added for this load.

christian-byrne commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,313 @@
+        )
+      const { subgraphs } = this.activeState.definitions
+      const { nodes } = this.activeState
+      //Instanceof doesn't funciton as nodes are serialized
```
nit: Typo in word "function"

christian-byrne commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,313 @@
+      const { nodes } = this.activeState
+      //Instanceof doesn't funciton as nodes are serialized
+      function isSubgraphNode(node: ComfyNode) {
+        return node && subgraphs.some((s) => s.id == node.type)
```
The loose equality `==` is intentional correct?

christian-byrne commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,313 @@
+  })
+}
+
+const subgraphCache: Record<string, LoadedComfyWorkflow> = {}
```
nit: Is it necessary for this to be in the module scope? The pinia store should also be a singleton but has benefit of being lazily loaded, better encapsulate for testing, accessible in vue dev tools. That said, this is also perfectly fine in my opinion.

christian-byrne commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,313 @@
+    settled
+      .filter((i) => 'reason' in i)
+      .forEach(({ reason }) =>
+        console.error('Failed to load subgraph blueprint', reason)
```
nit: Should we show a toast?

AustinMroz replied:
Agreed. I'm thinking always a single toast that either displays filenames, or number of failed loads if > ~3. Will keep the error message for detailed logging.

christian-byrne commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,313 @@
+    const canvas = canvasStore.getCanvas()
+    const subgraphNode = [...canvas.selectedItems][0]
+    if (
+      canvas.selectedItems.size != 1 ||
```
Is loose inequality intentional here?

christian-byrne commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,313 @@
+      canvas.selectedItems.size != 1 ||
+      !(subgraphNode instanceof SubgraphNode)
+    )
+      throw new TypeError('Must have single SubgraphNode selected to publish')
+
+    const { nodes = [], subgraphs = [] } = canvas._serializeItems([
+      subgraphNode
+    ])
+    if (nodes.length != 1) {
+      throw new TypeError('Must have single SubgraphNode selected to publish')
```
nit: Should we surface the errors to the user with toasts then return instead of throwing? Or is it necessary to throw?

christian-byrne commented on src/stores/subgraphStore.ts:
```diff
@@ -0,0 +1,313 @@
+    async function loadBlueprint(options: {
+      path: string
+      modified: number
+      size: number
```
nit: the semantics of the `size` field is a little unclear to me on initial read (might be my fault) and the semantics of `size: 1` are somewhat non-obvious maybe.

christian-byrne commented on src/components/sidebar/tabs/nodeLibrary/NodeTreeLeaf.vue:
```diff
@@ -80,6 +110,33 @@ const sidebarLocation = computed<'left' | 'right'>(() =>
 }
+const editBlueprint = async () => {
+  if (!props.node.data)
+    throw new Error(
+      'Failed to edit subgraph blueprint lacking backing node data'
+    )
```
nit: Would it be better to show a toast and return without throwing?

christian-byrne reviewed (changes requested):
This is looking great.

To help me, is this the correct flow?

1. User clicks "Publish Subgraph" button → publishSubgraph() called
2. Get selected subgraph node → Validates exactly 1 SubgraphNode selected
3. Serialize the node → canvas._serializeItems([subgraphNode])
4. Extract data structures → Gets nodes[] and subgraphs[] arrays
5. Create workflow JSON → Builds a complete workflow structure
6. Save as file → Stores in subgraphs/filename.json
7. Register as node type → Makes it available in node library
8. Cache for editing → Stores loaded workflow for future edits

That is my understanding, want to make sure it's correct.

Also, I like the refactoring of the clipboard logic into distinct methods responsible for clipboard functionality and serialization individually. It makes me want to look into some further refactor to some sort of serialization service like

```typescript
export function useCanvasSerializationService() {
    function serializeItems(items: Iterable<Positionable>): ClipboardItems {
      // All serialization logic here
    }

    function serializeForClipboard(items: Iterable<Positionable>): void {
      const data = serializeItems(items)
      localStorage.setItem('litegrapheditor_clipboard', JSON.stringify(data))
    }

    function serializeForBlueprint(subgraphNode: SubgraphNode): WorkflowData {
      const data = serializeItems([subgraphNode])
      return formatAsWorkflow(data)
    }

    return { serializeItems, serializeForClipboard, serializeForBlueprint }
}
```

Just thinking out loud. Also, feel free to ignore the nits.

AustinMroz committed:
Fix multiple nits

AustinMroz committed:
Further cleanup: error handling, and comments

AustinMroz committed:
Fixing failing test cases

This also moves the bg tinting to a property of the workflow,
which makes things more extensible in the future.

AustinMroz committed:
Fix temporary marking on publish.

The prior fix to allow overwrite of an existing blueprint on publish was
misguided. By marking a not-yet-loaded file as non-temporary, the load
performed prior to saving was actually fetching the file off disk and
discarding the existing changes. This additionally entirely prevented
publishing when a blueprint did not already exist with the current name.

To fix this, the blueprint is not marked as non-temporary until after
the load occurs. Note that this load is still required as it initializes
the change tracker state required for saving.

AustinMroz committed:
Block unloading subgraph blueprints

Will need to be revisited if lazy loading is implemented, but this
requires solving some ugly sync/async issues.

AustinMroz committed:
Merge main into austin/subgraph-publishing

christian-byrne reviewed (approved):
LGTM!
